internal/dag/builder.go

```diff
@@ -787,6 +807,38 @@ func (b *Builder) computeRoutes(sw *ObjectStatusWriter, proxy *projcontour.HTTPP
 	return routes
 }
 
+func validateHeaderAlterations(alt *projcontour.HeaderAlterations) (map[string]string, []string, error) {
+	if alt == nil {
+		return nil, nil, nil
+	}
+	add := make(map[string]string)
+
+	for _, entry := range alt.Add {
+		key := http.CanonicalHeaderKey(entry.Name)
+		_, ok := add[key]
+		if ok {
+			return nil, nil, fmt.Errorf("Duplicate header addition: %q", key)
+		}
+		if msgs := validation.IsHTTPHeaderName(key); len(msgs) != 0 {
+			return nil, nil, fmt.Errorf("invalid add header %q: %v", key, msgs)
+		}
+		add[key] = entry.Value
```

Comment:
We should make an explicit choice here whether we support envoy [% encoded header values](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_conn_man/headers#custom-request-response-headers).

Reply:
Interesting, I was unaware of those capabilities.  I think my personal bias would be towards having `Value` in the API hold the value of a raw string that is passed through.  This doesn't preclude adding another mutually exclusive key in the struct later for allowing these variables (e.g. `ValueWithVariables`) if the desire presents itself.

I'm happy to go either way on this, but will bias towards escaping everything for now.  thanks for catching this!

Reply:
In #70, there's a use case for interpolating K8s metadata to generate a linkerd header. So there's definitely good cases for interpolation, just need to talk it through and get consensus on what we want to support.